### PR TITLE
Adding code that checks if the nav sd card is installed, and if it is…

### PR DESCRIPTION
**Pull request by aidandj**
Tossing this in a pull request. Its not ready yet (needs to be gated somehow for other firmwares) but I wanted it on a branch so I could continue to update my own code as things advance. Could be useful if a config file becomes an option in the future too.

… reverse the heading passed to google maps. Only usefull on certain firmwares (v56, and apparently v55.760)

### CODE REVIEW

File: mazda/main.cpp [hunk 1]
```diff
@@ -114,7 +117,27 @@ static void gps_thread_func(std::condition_variable& quitcv, std::mutex& quitmut
...
 
-                location->set_bearing(static_cast<int32_t>(data.heading * 1E6));
+                // If the sd card exists then reverse heading. This should only be used on installs that have the 
+                // reversed heading issue. Maybe gate it by a config file
+                const char* sdCardFolder;
+                sdCardFolder = SD_CARD_PATH;
+                struct stat sb;
+                double newHeading;
+    
+                if (stat(sdCardFolder, &sb) == 0 && S_ISDIR(sb.st_mode))
```
Comment (mishan) [inline mazda/main.cpp]: Although stat() is super cheap, how often does this method get called? Can we perhaps cache the existence / non-existence of the directory and just check a variable here instead?

Comment (aidandj) [reply mazda/main.cpp]: Every second. The problem with caching it is if you pull it out while running it will reverse. This code allows you to plug and unplug the sdcard while running google maps

Comment (mishan) [reply mazda/main.cpp]: I understand this -- is that really an important use case to cover, though? My understanding is that people will either have an SD card or not.

I think you're on the right track here, but I'd rather not add another syscall to perform every second. Can we check periodically and flip a flag if needed? A check every 30 seconds? People pulling out and putting their SD card back in seems very edge case-y. I know it's great for testing though :)

Comment (mishan) [reply mazda/main.cpp]: I mean, we could always just tell users they have to restart AA if they're going to pull out or re-insert their Mazda Nav SD card -- that doesn't sound unreasonable to me. And it makes the engineering solution simpler. My personal philosophy is that sometimes it's better to set clear product requirements than to engineer for edge cases, especially if it costs any performance

Comment (aidandj) [reply mazda/main.cpp]: That's definitely a better idea. I'd have to do some more investigating to do a timer.

Comment (aidandj) [reply mazda/main.cpp]: I'm fine either way. Honestly I'm never gonna pull the SD card out, I just figured I'd toss it in. I had done something similar at work recently so just threw it in. I'm not married to either idea, I really just did it for myself, and figured I'd contribute.

Comment (mishan) [reply mazda/main.cpp]: We definitely appreciate the contribution and would like to see the bearing check merged in. I've been watching all the PRs and issue comments from the sideline. Nothing wrong with stat() but I don't like calling it every second. 

You can definitely add some sort of timer into headunit to do a periodic check or "Ostrich Algorithm" (AKA "don't care to handle") and document it for the end-user. I think either of those would work out great.

Since I'm not the maintainer of this project, it's not my call to make, so I'm just giving my opinion here.

Comment (mishaaq) [reply mazda/main.cpp]: Frankly speaking, I insert and remove the nav card while driving only because of this issue 😄. I agree with @mishan that this is such a super edge case that the user could restart the AA after insert/remove.

Comment (aidandj) [reply mazda/main.cpp]: I agree with the above. I think it should just check on startup. I keep meaning to test a build without the check because my performance has not been great lately with google maps (but I didn't have a lot of experience before the issue, car is new to me.)

Comment (Trevelopment) [reply mazda/main.cpp]: Yea I did a test and it is the performance impact from that check is definitely noticeable.

Comment (aidandj) [reply mazda/main.cpp]: You have to wait for a bit for it to flip. But interesting.

I will install a version without the check soon. Did you push your changes to any branch?

Comment (Trevelopment) [reply mazda/main.cpp]: no but I only used lines 125, 129-133 & 140 thats why its so funny because at that point its just a forced mathematical change of +-180 but somehow it still knows how to correct itself.

Comment (Trevelopment) [reply mazda/main.cpp]: After 2 days still works perfectly for me without flipping the heading or any other unexpected issues with or without a Nav SD card connected.  Tested it on some other people too and nobody reported a reversed heading whether they had the issue to begin with or not.  Its very strange that it wouldn't flip the heading for those that didn't have the issue to begin with but so far thats what its looking like.  I would say if you just poll for the Nav SD card once at the start that would be sufficient and if after some more testing it is confirmed that it doesn't cause an issue for anyone then I would say merge it because it definitely fixes the issue for those who have it, that has pretty much been confirmed by several people.

Comment (Peck07) [reply mazda/main.cpp]: I've merged this code with my config file comibnation:
https://github.com/Peck07/headunit/commit/b26e717b2e2ce38ecd508c1ab47f8ff180092d84

**Comment by Bastien7** [comment]
It has been 6 months ago now, what about merging this pull request?
I have the exact problem that this pull request is supposed to solve.

**Comment by Peck07** [comment]
Bastien7: actually I think this merge request contans this change as well:
https://github.com/gartnera/headunit/pull/134/
configurable with the reverseGPS parameter

**Comment by Bastien7** [comment]
Thanks for your response, I checked the current code and yes you're right, the code proposal has been integrated even if the pull request has the status "opened".
